k8s/pkg/apis/longhorn/v1beta2/volumeattachment.go

```diff
@@ -0,0 +1,128 @@
+}
+
+// VolumeError captures an error encountered during a volume operation.
+type VolumeError struct {
+	// Time the error was encountered.
+	// +optional
+	Time metav1.Time `json:"time,omitempty"`
+
+	// String detailing the error encountered during Attach or Detach operation.
+	// This string may be logged, so it should not contain sensitive
+	// information.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+type AttacherType string
+
+const (
+	AttacherTypeCSIAttacher             = AttacherType("csi-attacher")
+	AttacherTypeLonghornAPI             = AttacherType("longhorn-api")
+	AttacherTypeSnapshotController      = AttacherType("snapshot-controller")
+	AttacherTypeBackupController        = AttacherType("backup-controller")
+	AttacherTypeCloningController       = AttacherType("cloning-controller")
+	AttacherTypeSalvageController       = AttacherType("salvage-controller")
+	AttacherTypeShareManagerController  = AttacherType("share-manager-controller")
+	AttacherTypeLiveMigrationController = AttacherType("live-migration-controller")
+	AttacherTypeLonghornUpgrader        = AttacherType("longhorn-upgrader")
+)
+
+const (
+	AttacherPriorityLevelLonghornAPI             = 1000
+	AttacherPriorityLevelCSIAttacher             = 900
+	AttacherPriorityLevelSalvageController       = 900
+	AttacherPriorityLevelShareManagerController  = 900
+	AttacherPriorityLevelLonghornUpgrader        = 900
+	AttacherPriorityLevelLiveMigrationController = 800
+	AttacherPriorityLevelSnapshotController      = 800
+	AttacherPriorityLevelBackupController        = 800
+	AttacherPriorityLevelCloningController       = 800
+)
+
+func GetAttacherPriorityLevel(t AttacherType) int {
+	switch t {
+	case AttacherTypeCSIAttacher:
+		return AttacherPriorityLevelCSIAttacher
+	case AttacherTypeLonghornAPI:
+		return AttacherPriorityLevelLonghornAPI
+	case AttacherTypeSnapshotController:
+		return AttacherPriorityLevelSnapshotController
+	case AttacherTypeBackupController:
+		return AttacherPriorityLevelBackupController
+	case AttacherTypeCloningController:
+		return AttacherPriorityLevelCloningController
+	case AttacherTypeSalvageController:
+		return AttacherPriorityLevelSalvageController
+	case AttacherTypeShareManagerController:
+		return AttacherPriorityLevelShareManagerController
+	case AttacherTypeLiveMigrationController:
+		return AttacherPriorityLevelLiveMigrationController
+	case AttacherTypeLonghornUpgrader:
+		return AttacherPriorityLevelLonghornUpgrader
+	default:
+		return 0
+	}
+}
+
+// VolumeAttachmentSpec defines the desired state of Longhorn VolumeAttachment
+type VolumeAttachmentSpec struct {
+	// +optional
+	Attachments map[string]*Attachment `json:"attachments"`
+	// The name of Longhorn volume of this VolumeAttachment
+	Volume string `json:"volume"`
+}
+
+// VolumeAttachmentStatus defines the observed state of Longhorn VolumeAttachment
+type VolumeAttachmentStatus struct {
+	// +optional
+	Attachments map[string]*Attachment `json:"attachments"`
```

Comment:
Could you please explain a little more?

Reply:
The spec does not need some fields like `Attached`, `AttachError` and `DetachError` is not required.
And as for the status part, `Parameters` and `Type` are not necessary.

Besides, with this, you don't need to create new structs in `api/model`

Reply:
> And as for the status part, Parameters and Type are not necessary.

I think it is necessary because we allow the attachment Parameters to be modified on the Spec. We need to record this information in status to see if the current attachment status still satisfies the attachment spec.

> The spec does not need some fields like Attached, AttachError and DetachError is not required.

Agree. But  Attached, AttachError and DetachError are pointer so they will not be shown in spec by default.

> Besides, with this, you don't need to create new structs in api/model

I am not very sure how we can achieve this. Could you elaborate more? 

Btw, the advantage of not separating them is that it is convenient to copy the spec to status directly